cosmwasm/packages/wasmi-runtime/src/wasm/io.rs

```diff
@@ -65,99 +66,89 @@ pub fn encrypt_output(
+            *ok = encrypt_serializeable(&key, ok)?;
         }
-    } else if v["Ok"].is_object() {
-        // init or handle or migrate
-        if let Value::Object(ok) = &mut v["Ok"] {
-            if ok["messages"].is_array() {
-                if let Value::Array(msgs) = &mut ok["messages"] {
-                    for msg in msgs {
-                        if msg["wasm"]["execute"]["msg"].is_string() {
-                            if let Value::String(msg_b64) = &mut msg["wasm"]["execute"]["msg"] {
-                                let mut msg_to_pass = SecretMessage::from_base64(
-                                    msg_b64.clone(),
-                                    nonce,
-                                    user_public_key,
-                                )?;
-
-                                msg_to_pass.encrypt_in_place()?;
-
-                                msg["wasm"]["execute"]["msg"] = encode(&msg_to_pass.to_slice());
-                            }
-                        } else if msg["wasm"]["instantiate"]["msg"].is_string() {
-                            if let Value::String(msg_b64) = &mut msg["wasm"]["instantiate"]["msg"] {
-                                let mut msg_to_pass = SecretMessage::from_base64(
-                                    msg_b64.clone(),
-                                    nonce,
-                                    user_public_key,
-                                )?;
-
-                                msg_to_pass.encrypt_in_place()?;
-
-                                msg["wasm"]["instantiate"]["msg"] = encode(&msg_to_pass.to_slice());
-                            }
-                        }
-                    }
+
+        // Output is an object
+        // Encrypt all Wasm messages (keeps Bank, Staking, etc.. as is)
+        WasmOutput::OkObject { ok } => {
+            for msg in &mut ok.messages {
+                if let CosmosMsg::Wasm(wasm_msg) = msg {
+                    encrypt_wasm_msg(wasm_msg, nonce, user_public_key)?;
                 }
             }
 
-            if ok["log"].is_array() {
-                if let Value::Array(events) = &mut ok["log"] {
-                    for e in events {
-                        if e["key"].is_string() {
-                            if let Value::String(k) = &mut e["key"] {
-                                e["key"] = encrypt_serializeable(&key, k)?;
-                            }
-                        }
-                        if e["value"].is_string() {
-                            if let Value::String(v) = &mut e["value"] {
-                                e["value"] = encrypt_serializeable(&key, v)?;
-                            }
-                        }
-                    }
-                }
+            // Encrypt all logs
+            for log in &mut ok.log {
+                log.key = encrypt_serializeable(&key, &log.key)?;
+                log.value = encrypt_serializeable(&key, &log.value)?;
             }
```

Comment:
I believe you don't need the `&mut` in `&mut ok.log` or the `&`s in `&log.key` and `&log.value`

Reply:
`&mut ok.log` is essential because I change `log`'s value.
The other `&`s are because `encrypt_serializeable` takes `&T`

Reply:
does it complain if you remove them?
`ok` here is already a `&mut T`, so so should `ok.log`, `log.key`, and `log.value` be.
And `&mut` i usable as a `&`

Reply:
Yes it complains. It seems like the `for` statement loses the `&mut` on `ok`, maybe because I refer to the inner `ok.log`?

Reply:
hmm... might be.
the compiler has a bunch of specific behaviours in these contexts, so i was hoping it would be smart enought to understand these things. oh well :P